UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs

```diff
@@ -0,0 +1,209 @@
+                var options = new LibGit2Sharp.CheckoutOptions();
+                repository.Reset(LibGit2Sharp.ResetMode.Hard, (LibGit2Sharp.Commit)repository.Lookup(_commitHash, LibGit2Sharp.ObjectType.Commit), options);
+                repository.RemoveUntrackedFiles();
+            }
+
+            CommitHelper.SetCommitMessage(_moduleTestHelper.Module, commitMessageText: null, amendCommit: false);
+
+            _commands = new GitUICommands(_moduleTestHelper.Module);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+        }
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            _moduleTestHelper.Dispose();
+        }
+
+        [Test]
+        public void PreserveCommitMessageOnReopen()
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.IsEmpty(formCommit.GetTestAccessor().Message.Text);
+                formCommit.GetTestAccessor().Message.Text = generatedCommitMessage;
+            });
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.AreEqual(generatedCommitMessage, formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [TestCase(CommitKind.Fixup)]
+        [TestCase(CommitKind.Squash)]
+        public void DoNotPreserveCommitMessageOnReopenFromSpecialCommit(CommitKind commitKind)
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(
+                formCommit =>
+                {
+                    string prefix = commitKind.ToString().ToLowerInvariant();
+                    Assert.AreEqual($"{prefix}! A commit message", formCommit.GetTestAccessor().Message.Text);
+                    formCommit.GetTestAccessor().Message.Text = generatedCommitMessage;
+                },
+                commitKind);
+
+            RunFormCommitTest(formCommit =>
+            {
+                Assert.IsEmpty(formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        [Test]
+        public void SelectMessageFromHistory()
+        {
+            var generatedCommitMessage = Path.GetRandomFileName();
+
+            RunFormCommitTest(formCommit =>
+            {
+                var commitMessageToolStripMenuItem = formCommit.GetTestAccessor().CommitMessageToolStripMenuItem;
+
+                // Verify the message appears correctly
+                commitMessageToolStripMenuItem.ShowDropDown();
+                Assert.AreEqual("A commit message...", commitMessageToolStripMenuItem.DropDownItems[0].Text);
+
+                // Verify the message is selected correctly
+                commitMessageToolStripMenuItem.DropDownItems[0].PerformClick();
+                Assert.AreEqual("A commit message", formCommit.GetTestAccessor().Message.Text);
+            });
+        }
+
+        private static async Task WaitForIdleAsync()
```

Comment:
💭 I would consider this a candidate for refactoring once we establish some initial patterns.